### ISSUE TITLE
After switching from vstest v1 to v2, ASCII bell character in test causes vsts-agent to fail to publish results.

### ISSUE THREAD

**Issue by GaTechThomas** [Microsoft/vstest#683]
After opening this in the vsts-agent repo, it occurred to me that this is probably something that should be handled in the new version of vstest:

https://github.com/Microsoft/vsts-agent/issues/899

I apologize that I don't know the appropriate way to handle this cross post.

**Comment by TingluoHuang** [comment]
@nigurr

**Comment by codito** [comment]
@GaTechThomas thanks for raising this issue.

Error message that shows up is `"Failed to publish test results: '�', hexadecimal value 0x07, is an invalid character."`. @Faizan2304 is this handled in the work we did for escaping special characters (for invalid xml)?

**Comment by Faizan2304** [comment]
@codito : Yes we have fixed it but fix is in Visual Studio 2017. 
In vstest v1 they were taking care of invalid xml character when they read trx file but in vstest v2 it seems they are not. I talked with @nigurr and they want to go with this flow only and expecting trx file generator to give sanitized trx file.

**Comment by Faizan2304** [comment]
@GaTechThomas : I am not able to repro the issue with given repro test method shared in bug https://github.com/Microsoft/vsts-tasks/issues/3932 . I used below test method and got the repro

``` cSharp
public static TheoryData TestValues
            => new TheoryData<IEnumerable<KeyValuePair<string, string>>>
            {
                { new []{ new KeyValuePair<string,string>("key", "string with \a in it") } },
                { new []{ new KeyValuePair<string,string>("key2", "second string with \0 in it") } }
            };

        [Theory]
        [MemberData(nameof(TestValues))]
        public void Test2(IEnumerable<KeyValuePair<string, string>> values)
        {
        }
```
We have sanitized the invalid xml character in trx file in Visual studio 2017. I have verified and issue is not getting repro if we are generating trx with visual studio 2017.

**Comment by GaTechThomas** [comment]
Thank you. I have been testing VS2017, but not yet on VSTS. It has a few bugs that our team will have to wait before we can switch. However, I may be able to build on the server with VS2017. Probably won't be able to prioritize it for this week.


-Thomas


________________________________

**Comment by Faizan2304** [comment]
@GaTechThomas 
To get you unblock we need to have fix in two places:
1) Fix in TestPlatform side ( i'e vstest.console.exe, the one who is creating trx file) 
2) Fix in TestAgent (who is reading trx file and uploading test).

We have made fix in TestPlatform side in `VS 2017` but to consume fix you need pass custom vstest.console.exe path which is `VS-Insallation-Directory\Common7\IDE\Extensions\TestPlatform\vstest.console.exe` in your vstest V2  task.

@nigurr Can you please a bug on agent side and add link here for the same.

I m closing this bug from our side as we don't have any action item here.

**Comment by santoorman** [comment]
when i use this version of vstest.console.exe I get a bunch of UI unit test failures that I don't get with the default one used.  I am stuck because the tests pass with the default version, but that has the exception described in this issue when trying to publish results, this one will publish but these tests fail... any reason why these are failing in the version of vstest.console.exe in this folder: VS-Insallation-Directory\Common7\IDE\Extensions\TestPlatform\vstest.console.exe?

Test method .. threw exception:  System.InvalidOperationException: The calling thread must be STA, because many UI components require this.

**Comment by codito** [comment]
@santoorman the support for STA threads in `Extensions\TestPlatform\vstest.console.exe` was recently added (PR: https://github.com/Microsoft/vstest/pull/969). Documentation for this is at https://github.com/Microsoft/vstest-docs/blob/master/docs/configure.md#execution-thread-apartment-state.

This fix will ship as part of VS 2017.5 releases.
